src/io_cs.c

```diff
@@ -178,6 +183,139 @@ static int32_t io_cs_validate(
 }
 
 //
+// Build a unique key name from this connection string.
+//
+static int32_t io_cs_get_unique_key_name(
+    io_cs_t* cs,
+    STRING_HANDLE* created
+)
+{   
+    // The format is <policy>|device:<device>@<hostname>
+    int32_t result;
+    const char* id;
+    STRING_HANDLE key_name;
+
+    dbg_assert_ptr(cs);
+    dbg_assert_ptr(created);
+
+    // Get a unique key name as name/id @ hostname
+    id = io_cs_get_shared_access_key_name(cs);
+    key_name = STRING_construct(id ? id : "");
+    if (!key_name)
+        return er_out_of_memory;
+    do
+    {
+        result = er_out_of_memory;
+        if (!id)
+        {
+            id = io_cs_get_device_id(cs);
+            if (!id)
+            {
+                log_error(NULL, "No device nor key name to make "
+                    "unique key name for connection string!");
+                result = er_not_found;
+                break;
+            }
+
+            if (0 != STRING_concat(key_name, "device:") ||
+                0 != STRING_concat(key_name, id))
+                break;
+        }
+
+        id = io_cs_get_host_name(cs);
+        if (id &&
+            (0 != STRING_concat(key_name, "@") ||
+                0 != STRING_concat(key_name, id)))
+            break;
+        
+        *created = key_name;
+        return er_ok;
+    } 
+    while (0);
+    
+    STRING_delete(key_name);
+    return result;
+}
+
+//
+// If connection string contains key, protect it
+//
+static int32_t io_cs_import_key(
+    io_cs_t* cs
+)
+{
+    int32_t result;
+    size_t key_len;
+    STRING_HANDLE key_val;
+    STRING_HANDLE key_name = NULL;
+    STRING_HANDLE key_handle;
+    uint8_t* key = NULL;
+    bool persist;
+
+    // Key present?  And import supported?
+    key_val = cs->entries[io_cs_entry_shared_access_key];
+    if (!key_val || 0 == (pal_caps() & pal_cap_cred))
+        return er_ok;
+
+    result = io_cs_get_unique_key_name(cs, &key_name);
+    if (result != er_ok)
+        return result;
+
+    // base 64 decode the key then import into credential store  
+    // TODO: see if we need utf8!
+    do
+    {
+        result = string_base64_to_byte_array(STRING_c_str(key_val),
+            &key, &key_len);
+        if (result != er_ok)
+            break;
+
+        // Do not persist policy keys
+        persist = !io_cs_get_shared_access_key_name(cs) ||
+            __prx_config_get_int(prx_config_key_policy_import, 0);
+        result = pal_cred_import(STRING_c_str(key_name), key, key_len,
+            persist, &key_handle);
+        if (result != er_ok)
+            break;
+
+        // Success, delete original key
+        if (0 != STRING_compare(key_name, key_handle))
+            cs->entries[io_cs_entry_shared_access_key_handle] = key_handle;
+        else // No need to keep it - it is a unique well known key-name
+            STRING_delete(key_handle);
+
+        // Clear key value. 
+        // TODO: Need secure storage
```

Comment:
Ha, I must have been someplace else when I wrote this comment.  I meant secure **memory**, i.e. non-pageable.  I do not think it is a hi prio threat on import path, but we might want to add a backlog item.